### DISCUSSION TITLE
HarmonicAverage

### DISCUSSION THREAD

**Discussion by heinono1** [idaholab/moose#22306] [Q&A General] [answered]
It looks like HarmonicAverage is not a member of the Moose::FV::InterpMethod although there are examples (https://mooseframework.inl.gov/syntax/FVKernels/) that include HarmonicAverage as member of it. Will HarmonicAverage be added to that namespace?

**Comment by GiudGiud** [comment, marked as answer]
Hello

It is, did you update to the latest moose? We only added this recently. 

In FVDiffusion you can see:
```
  const auto & interp_method = getParam<MooseEnum>("coeff_interp_method");
  if (interp_method == "average")
    _coeff_interp_method = Moose::FV::InterpMethod::Average;
  else if (interp_method == "harmonic")
    _coeff_interp_method = Moose::FV::InterpMethod::HarmonicAverage;
```

**Reply by heinono1** [reply]
Ah, OK - thanks!

**Reply by heinono1** [reply]
It's weird (to me, at least): if I add "harmonic" to the interp_method, the code throws and error:  Invalid option "harmonic" in MooseEnum. Valid options (not case-sensitive) are "average"". What is weird is that I pulled and build a new distribution and there are FV kernels like FVDiffusion that include "harmonic" as an option to interp_method. I use exactly the same syntax as in, eg FVDiffusion or FVAnisotropicDiffusion: in validParams I have
```
MooseEnum coeff_interp_method("average harmonic","average");
paras.addParam<MooseEnum>("coeff_interp_method", coeff_interp_method, "Switch to select interpolation method");
```
That seems to be what throws the error. If I change to 
```
MooseEnum coeff_interp_method("average");
paras.addParam<MooseEnum>("coeff_interp_method", coeff_interp_method, "Switch to select interpolation method");
````
then I do not get the error. What is even weirder is that the error shows up even when I do not include this particular kernel in the input file.

**Reply by GiudGiud** [reply]
the second one will definitely not allow harmonic. 
Could you please paste the input file syntax that fails with the first one?

`params` is mispelled as paras btw.
can you please paste the entire validParams block for your object?

**Comment by heinono1** [comment]
Is TAO also added to the main branch or is it still in a develop branch?

On Wed, Oct 5, 2022 at 8:30 AM Guillaume Giudicelli <
***@***.***> wrote:

> Hello
>
> It is, did you update to the latest moose? We only added this recently.
>
> In FVDiffusion you can see:
>
>   const auto & interp_method = getParam<MooseEnum>("coeff_interp_method");
>   if (interp_method == "average")
>     _coeff_interp_method = Moose::FV::InterpMethod::Average;
>   else if (interp_method == "harmonic")
>     _coeff_interp_method = Moose::FV::InterpMethod::HarmonicAverage;
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/discussions/22306#discussioncomment-3805268>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEKZEF3D3DW7ZW4GJZ7SSV3WBV7H3ANCNFSM6AAAAAAQ5SRO3Q>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Olle Heinonen
***@***.***

**Reply by GiudGiud** [reply]
Still in this branch:
https://github.com/idaholab/moose/pull/21887

if you need both you should be able to rebase it on top of devel

**Comment by heinono1** [comment]
Cool - thanks!

On Wed, Oct 5, 2022 at 9:04 AM Guillaume Giudicelli <
***@***.***> wrote:

> Still in this branch:
> #21887 <https://github.com/idaholab/moose/pull/21887>
>
> if you need both you should be able to rebase it on top of devel
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/discussions/22306#discussioncomment-3805563>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEKZEF2KEQFKPCWRQ2JXYHDWBWDFXANCNFSM6AAAAAAQ5SRO3Q>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Olle Heinonen
***@***.***

**Comment by heinono1** [comment]
Hi Guillaume,

Below are the relevant code lines. When I have commented out the lines with
"harmonic", it works. When I put those lines back in (and comment out the
lines with just "average"), execution throws an error, even if I do not use
this kernel in the execution.

```
InputParameters
FVMagneticExchangeNonCons::validParams()
{
  InputParameters params = FVFluxKernel::validParams();
  params.addClassDescription("Computes residual of conservative exchange
torque");
  params.addRequiredParam<unsigned int>("component", "An integer
corresponding to the direction in order parameter space");
  params.addRequiredCoupledVar("mag_x", "The x component of the constrained
magnetic vector");
  params.addRequiredCoupledVar("mag_y", "The y component of the constrained
magnetic vector");
  params.addRequiredCoupledVar("mag_z", "The z component of the constrained
magnetic vector");
  MooseEnum coeff_interp_method("average");
  //  MooseEnum coeff_interp_method("average harmonic", "harmonic");
  params.addParam<MooseEnum>(
    "coeff_interp_method",
    coeff_interp_method,
    "Switch that can select interpolation method for magnetization
density.");
  params.set<unsigned short>("ghost_layers") = 2;
  return params;
}
```

```
FVMagneticExchangeNonCons::FVMagneticExchangeNonCons(const InputParameters
& params)
  : FVFluxKernel(params), /*_coeff(getFunctor<ADReal>("coeff")),*/
   _component(getParam<unsigned int>("component")),
   _mag_x_var(coupled("mag_x")),
   _mag_y_var(coupled("mag_y")),
   _mag_z_var(coupled("mag_z")),
   _mag_x(coupledValue("mag_x")),
   _mag_y(coupledValue("mag_y")),
   _mag_z(coupledValue("mag_z")),
   _mag_x_grad(coupledGradient("mag_x")),
   _mag_y_grad(coupledGradient("mag_y")),
   _mag_z_grad(coupledGradient("mag_z")),
   _alpha(getFunctor<ADReal>("alpha")),
   _g0(getFunctor<ADReal>("g0")),
   _Ms(getFunctor<ADReal>("Ms")),
   _Ae(getFunctor<ADReal>("Ae"))
{
#ifndef MOOSE_GLOBAL_AD_INDEXING
  mooseError(
      "FVMagneticExchangeCons is not supported by local AD indexing. In
order to use this object, please run "
      "the configure script in the root MOOSE directory with the configure
option "
      "'--with-ad-indexing-type=global'. Note that global indexing is now
the default "
      "configuration for AD indexing type.");
#endif

  const auto & interp_method = getParam<MooseEnum>("coeff_interp_method");
  if (interp_method == "average")
    const auto _coeff_interp_method = Moose::FV::InterpMethod::Average;
  /*  else if (interp_method == "harmonic")
      _coeff_interp_method = Moose::FV::InterpMethod::HarmonicAverage;*/

  if ((_var.faceInterpolationMethod() ==
Moose::FV::InterpMethod::SkewCorrectedAverage) &&
      (_tid == 0))
     adjustRMGhostLayers(std::max((unsigned short)(3), _pars.get<unsigned
short>("ghost_layers")));
}
```

**Reply by GiudGiud** [reply]
I just copy pasted this all into FVDiffusion and it ran fine. 
except for this:

``
    const auto _coeff_interp_method = Moose::FV::InterpMethod::Average;
``
this is not what we want. _coeff_interp_method is a class attribute its type should not be declared in the source but in the header

If you update MOOSE again, I just add a selectInterpolationMethod routine that does this for you. 
you call it like in FVDiffusion.C
```
    _coeff_interp_method(
        Moose::FV::selectInterpolationMethod(getParam<MooseEnum>("coeff_interp_method")))
```

**Comment by heinono1** [comment]
Thanks, Guillaume.

On Wed, Oct 19, 2022 at 6:15 PM Guillaume Giudicelli <
***@***.***> wrote:

> I just copy pasted this all into FVDiffusion and it ran fine.
> except for this:
>
> const auto _coeff_interp_method = Moose::FV::InterpMethod::Average;
> this is not what we want. _coeff_interp_method is a class attribute its
> type should not be declared in the source but in the header
>
> If you update MOOSE again, I just add a selectInterpolationMethod routine
> that does this for you.
> you call it like in FVDiffusion.C
>
>     _coeff_interp_method(
>         Moose::FV::selectInterpolationMethod(getParam<MooseEnum>("coeff_interp_method")))
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/discussions/22306#discussioncomment-3919653>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEKZEF5JT4H4G2KRJLJ4HF3WEB6KRANCNFSM6AAAAAAQ5SRO3Q>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
Olle Heinonen
***@***.***
